runtime/lua/vim/filetype/detect.lua

```diff
@@ -869,6 +869,67 @@ function M.y(bufnr)
   vim.bo[bufnr].filetype = 'yacc'
 end
 
+function M.decl(bufnr)
+  for _, line in ipairs(getlines(bufnr, 1, 3)) do
+    if line:lower():find('^<!sgml') then
+      vim.bo[bufnr].filetype = 'sgmldecl'
+      return
+    end
+  end
+end
+
+function M.dcl(bufnr)
+  for _, line in ipairs(getlines(bufnr, 1, 3)) do
+    if line:lower():find('^<!sgml') then
+      vim.bo[bufnr].filetype = 'sgmldecl'
+      return
+    end
+  end
+  vim.bo[bufnr].filetype = 'clean'
+end
+
+function M.sgml(bufnr)
+  if getlines(bufnr, 1, 5, { concat = true }):lower():find('linuxdoc') then
+    vim.bo[bufnr] = 'smgllnx'
+    return
+  end
+  if getlines(bufnr, 1, 5, { concat = true }):lower():find('<!DOCTYPE.*DocBook') then
```

Comment:
No `:lower()` needed here, also I would replace the `return` with an else statement